test/Semantics/assign07.f90

```diff
@@ -0,0 +1,35 @@
+! RUN: %B/test/Semantics/test_errors.sh %s %flang %t
+! Test ASSIGN statement, assigned GOTO, and assigned format labels
+! (see subclause 8.2.4 in Fortran 90 (*not* 2018!)
+
+program main
+  call test(0)
+2 format('no')
+ contains
+  subroutine test(n)
+    !ERROR: Label '4' is not a branch target or FORMAT
+4   integer, intent(in) :: n
+    integer :: lab
+    assign 1 to lab ! ok
+    assign 1 to implicitlab1 ! ok
+    !ERROR: Label '666' was not found
+    assign 666 to lab
+    !ERROR: Label '2' was not found
+    assign 2 to lab
+    assign 4 to lab
+    if (n==1) goto lab ! ok
+    if (n==1) goto implicitlab2 ! ok
+    if (n==1) goto lab(1) ! ok
```

Comment:
You could have an example where the label has not been assigned to any one of the list of statement labels.
```
assign 10 to lab2
if (n==1) goto lab2(1,2)
.
.
10 <some executable instruction>
1 <another executable instruction>
2 <a third executable instruction>

```

Reply:
This should produce an error.

Reply:
Another test case. This should also produce an error.
```
assign 1 to lab
lab2 = lab
goto lab2
```

Reply:
Why are these errors?

Reply:
I don't have the Fortran 90 doc with me but I found elsewhere these rules
```
- If a list of labels appears, the statement label assigned to the variable must be one of the labels in the list.
- The variable must have a statement label value assigned to it by an ASSIGN statement (not an arithmetic assignment statement) before the GO TO statement is executed.
```

Reply:
Sure, but neither of these restrictions imposed on the programmer can be detected reliably as errors at compilation time without data-flow analysis.

Reply:
I too interpret these two cases as violating program requirements.
I find quite a lot of variation in both error checking and run time behavior across other compilers.

Reply:
I can agree with that.

Reply:
Single pass static analysis should suffice to diagnose these two errors.  (Expected code generation is to have a map of symbols --> assign sets.)

Reply:
They do violate requirements on the program.  But I don't think that we'll be able to emit compile-time errors reliably without data-flow analysis, and we don't have a CFG until semantics; the best we could do would be warnings.  As these are deleted features, I don't want to spend a whole lot of programmer or compilation time on them.

Reply:
A simple scan would catch obvious violations such as these two cases.  However, these are program requirements, not constraint violations, and helping a programmer or porter to use this functionality doesn't seem much of a priority, even in comparison to other legacy functionality.